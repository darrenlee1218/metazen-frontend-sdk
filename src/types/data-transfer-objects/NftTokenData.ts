/**
 * @address address of current wallet
 * @assetId chainId_ticker_0x${contractAddress.substring(0, 8)}
 * @nftTokenId tokenId in the ERC721 / ERC1155 contract
 * @assetTicker deprecated
 */
interface NftTokenData {
  entityId: string;
  chainId: string;
  address: string;
  balance: string;
  assetId: string;
  nftTokenId: string;
  assetTicker: string;
}

export default NftTokenData;
