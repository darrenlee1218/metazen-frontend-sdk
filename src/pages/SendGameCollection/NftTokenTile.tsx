import React, { FC, useCallback, useMemo } from 'react';

import { CustomTheme, useTheme } from '@mui/material/styles';
import { AssetVariant, NftAssetTile } from '@components/NftAssetTile';
import { useGetAssetMetadataQuery } from '@redux/api/assetMetadataCache';
import NftTokenData from '@gryfyn-types/data-transfer-objects/NftTokenData';

interface NftTokenTileProps {
  nftToken: NftTokenData;
  collectionContractAddress: string;
  isSelected: boolean;
  onClick: (tokenKey: string | null) => void;
}

export const NftTokenTile: FC<NftTokenTileProps> = ({ nftToken, collectionContractAddress, isSelected, onClick }) => {
  const theme = useTheme() as CustomTheme;
  const { chainId, nftTokenId } = nftToken;
  const { data: nftMetadata } = useGetAssetMetadataQuery({
    chainId,
    contractAddress: collectionContractAddress,
    tokenId: nftTokenId,
  });

  const { assetIconUrl, assetName } = useMemo(() => {
    if (nftMetadata?.data) {
      return {
        assetIconUrl: nftMetadata.data.image,
        assetName: nftMetadata.data.name,
      };
    }

    return {
      assetIconUrl: '',
      assetName: '',
    };
  }, [nftMetadata]);

  const handleClick = useCallback(() => {
    onClick(isSelected ? null : nftTokenId ?? null);
  }, [isSelected, nftTokenId, onClick]);

  return (
    <NftAssetTile
      variant={AssetVariant.Collection}
      assetIconUrl={assetIconUrl}
      assetName={assetName}
      assetQuantity={null}
      onClick={handleClick}
      sx={{
        border: `1px solid ${isSelected ? theme.palette.primary.main : theme.palette.colors.border}`,
        ':hover': {
          borderColor: isSelected ? theme.palette.primary.main : theme.palette.colors.clickableGrayHover,
        },
      }}
    />
  );
};
