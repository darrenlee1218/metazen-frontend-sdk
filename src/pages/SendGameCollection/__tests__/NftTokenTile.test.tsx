import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import NftTokenData from '@gryfyn-types/data-transfer-objects/NftTokenData';
import { NftTokenTile } from '../NftTokenTile';
import { ThemeProvider } from '@mui/material/styles';
import defaultTheme from '@theme/theme';

const mockNftToken = {} as unknown as NftTokenData;

jest.mock('@redux/api/assetMetadataCache', () => ({
  __esModule: true,
  useGetAssetMetadataQuery: () => ({
    data: {
      data: {
        name: 'Nft Name',
        image: '404.png',
      },
    },
    isLoading: false,
  }),
}));

describe('NftTokenTile', () => {
  it('should render correctly', () => {
    const handleClick = jest.fn();
    const baseDom = render(
      <ThemeProvider theme={defaultTheme()}>
        <NftTokenTile isSelected collectionContractAddress="" nftToken={mockNftToken} onClick={handleClick} />
      </ThemeProvider>,
    );
    expect(baseDom.baseElement).toMatchSnapshot();
  });

  it('should call the clickHandler on user click', () => {
    const handleClick = jest.fn();
    const baseDom = render(
      <ThemeProvider theme={defaultTheme()}>
        <NftTokenTile isSelected collectionContractAddress="" nftToken={mockNftToken} onClick={handleClick} />
      </ThemeProvider>,
    );
    fireEvent.click(baseDom.getByText('Nft Name'));
    expect(handleClick).toHaveBeenCalled();
  });
});
