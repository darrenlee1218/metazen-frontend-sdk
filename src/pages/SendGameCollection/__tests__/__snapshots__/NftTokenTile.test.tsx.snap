// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NftTokenTile should render correctly 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiCardActionArea-root MuiBox-root css-evq3ry-MuiButtonBase-root-MuiCardActionArea-root"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiBox-root css-hjbv5f"
      >
        <img
          alt="Nft Name image"
          class="MuiBox-root css-djl8wu"
          src="404.png"
        />
      </div>
      <div
        class="MuiBox-root css-120alf"
        data-testid="asset-descriptor"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-11hf9u5-MuiTypography-root"
        >
          Nft Name
        </h5>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
      />
    </button>
  </div>
</body>
`;
